File: apps/askbible-android/app/src/main/java/me/askbible/native_/ui/VerseActionSheet.kt
package me.askbible.native_.ui

import me.askbible.native_.data.SiteCopy
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import me.askbible.native_.data.Parchment
import me.askbible.native_.data.ReadSize

/**
 * 长按某节弹出的操作单。对应 RN ReadChapterScreenVerseActionModal（ReadChapterBottomSheet 外壳），与 iOS 的 VerseActionSheet 对等：
 * 标题「第 N 节」+ 右上「关闭」，三列格子：本节复制 / 双击收藏（已收藏则不出）/ 分享。RN 还有「多选复制」「划重点」两项，本轮未接。
 */
@Composable
fun VerseActionSheet(
    verse: Int,
    bookmarked: Boolean,
    size: ReadSize,
    onCopy: () -> Unit,
    onBookmark: () -> Unit,
    onShare: () -> Unit,
    onClose: () -> Unit,
    theme: Parchment = Parchment.light,
) {
    val fs = size.metrics.verseFontSize
    val iconSize = maxOf(22f, Math.round(fs * 1.15f).toFloat())
    val labelSize = maxOf(13f, Math.round(fs * 0.78f).toFloat())
    Box(Modifier.fillMaxSize().background(Color(0x591C1410)).clickableNoRipple(onClose), contentAlignment = Alignment.BottomCenter) {
        Column(
            Modifier.fillMaxWidth().navigationBarsPadding().parchmentCard(16.dp)
                .padding(start = 18.dp, end = 18.dp, top = 14.dp, bottom = 28.dp).clickableNoRipple {},
        ) {
            Row(Modifier.fillMaxWidth().padding(bottom = 10.dp), verticalAlignment = Alignment.Top, horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                Text("第 $verse 节", Modifier.weight(1f), color = theme.ink.toColor(), fontSize = maxOf(17f, Math.round(fs * 0.95f).toFloat()).sp, fontWeight = FontWeight.SemiBold)
                Text("关闭", Modifier.clickableNoRipple(onClose), color = theme.muted.toColor(), fontSize = Math.round(fs * 0.85f).sp)
            }
            Row(Modifier.fillMaxWidth().padding(top = 4.dp, bottom = 8.dp), verticalAlignment = Alignment.Top) {
                Cell(MI.CONTENT_COPY, SiteCopy.t("pages.read.verseActionCopy"), iconSize, labelSize, theme, Modifier.weight(1f), onCopy)
                if (!bookmarked) Cell(MI.BOOKMARK_BORDER, SiteCopy.t("pages.read.verseActionBookmark"), iconSize, labelSize, theme, Modifier.weight(1f), onBookmark)
                Cell(MI.IOS_SHARE, SiteCopy.t("pages.read.verseActionShare"), iconSize, labelSize, theme, Modifier.weight(1f), onShare)
            }
        }
    }
}

@Composable
private fun Cell(glyph: String, label: String, icon: Float, font: Float, theme: Parchment, modifier: Modifier, onClick: () -> Unit) {
    Column(modifier.clickableNoRipple(onClick).padding(horizontal = 6.dp, vertical = 14.dp),
           horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        MaterialIcon(glyph, icon, theme.ink.toColor())
        Text(label, color = theme.ink.toColor(), fontSize = font.sp, lineHeight = 18.sp, fontWeight = FontWeight.Medium, textAlign = TextAlign.Center, maxLines = 2)
    }
}

/** 收藏 / 复制后的轻提示（RN ReadVerseBookmarkFeedback）：深色胶囊，14/500 #f7f4ef */
@Composable
fun VerseFeedbackToast(message: String, modifier: Modifier = Modifier) {
    Box(modifier.background(Color(0xD11C1410), CircleShape).padding(horizontal = 16.dp, vertical = 10.dp)) {
        Text(message, color = Color(0xFFF7F4EF), fontSize = 14.sp, fontWeight = FontWeight.Medium, textAlign = TextAlign.Center)
    }
}

/**
 * 复制 / 分享用的经文文案，与 RN 同一格式：
 * 复制 = 「书名 章:节 经文」一行（lib formatScriptureVerseClipboard）；分享 = 「书名 章:节」换行经文（Share.share message）
 */
object VerseShareText {
    fun clipboard(bookName: String, chapter: Int, verse: Int, text: String) = "$bookName $chapter:$verse ${text.trim()}".trim()
    fun share(bookName: String, chapter: Int, verse: Int, text: String) = "$bookName $chapter:$verse\n$text"
}

File: apps/askbible-ios/AskBible/Read/VerseActionSheet.swift
import SwiftUI
import UIKit

/// 长按某节弹出的操作单。对应 RN ReadChapterScreenVerseActionModal（ReadChapterBottomSheet 外壳）：
/// 标题「第 N 节」+ 右上「关闭」，三列格子：本节复制 / 双击收藏（已收藏则不出）/ 分享。
/// RN 的「划重点」还没接；「多选复制」2026-09-11 补上了。
struct VerseActionSheet: View {
    let verse: Int
    let bookmarked: Bool
    var size: ReadSize = .default
    var onCopy: () -> Void
    var onBookmark: () -> Void
    var onShare: () -> Void
    /// 多选复制：以这一节为起点进入选择态（RN runStartMultiCopy）
    var onMultiCopy: () -> Void = {}
    /// 划重点：进入划字模式（RN runOpenHighlightEditor）
    var onHighlight: () -> Void = {}
    var onClose: () -> Void

    private let theme = Parchment.light

    var body: some View {
        let fs = size.metrics.verseFontSize
        let iconSize = max(22, (fs * 1.15).rounded())
        let labelSize = max(13, (fs * 0.78).rounded())
        ZStack(alignment: .bottom) {
            Color(red: 28 / 255, green: 20 / 255, blue: 16 / 255, opacity: 0.35)
                .frame(maxWidth: .infinity, maxHeight: .infinity)
                .ignoresSafeArea()
            VStack(alignment: .leading, spacing: 0) {
                HStack(alignment: .top, spacing: 12) {
                    Text(SiteCopy.f("pages.read.verseActionVerseTitle", ["verse": "\(verse)"])).font(.system(size: max(17, (fs * 0.95).rounded()), weight: .semibold)).foregroundStyle(theme.ink)
                        .frame(maxWidth: .infinity, alignment: .leading)
                    Button(action: onClose) { Text(SiteCopy.t("pages.read.verseActionClose")).font(.system(size: (fs * 0.85).rounded())).foregroundStyle(theme.muted) }
                        .buttonStyle(.plain)
                }
                .padding(.bottom, 10)
                HStack(alignment: .top, spacing: 0) {
                    cell(MI.contentCopy, SiteCopy.t("pages.read.verseActionCopy"), iconSize, labelSize, onCopy)
                    if !bookmarked { cell(MI.bookmarkBorder, SiteCopy.t("pages.read.verseActionBookmark"), iconSize, labelSize, onBookmark) }
                    cell(MI.iosShare, SiteCopy.t("pages.read.verseActionShare"), iconSize, labelSize, onShare)
                    cell(MI.libraryAddCheck, SiteCopy.t("native.verseMultiCopy"), iconSize, labelSize, onMultiCopy)
                    cell(MI.brush, SiteCopy.t("native.highlight"), iconSize, labelSize, onHighlight)
                }
                .padding(.top, 4).padding(.bottom, 8)
            }
            .padding(.horizontal, 18).padding(.top, 14).padding(.bottom, 28)
            .frame(maxWidth: .infinity, alignment: .leading)
            .parchmentCard(cornerRadius: 16)
            .contentShape(Rectangle())
            .onTapGesture {}
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        // 点空白关面板：手势挂整层（见 TranslationPanel 那条注）
        .contentShape(Rectangle())
        .onTapGesture(perform: onClose)
    }

    private func cell(_ glyph: String, _ label: String, _ icon: CGFloat, _ font: CGFloat, _ action: @escaping () -> Void) -> some View {
        Button(action: action) {
            VStack(spacing: 8) {
                MaterialIcon(glyph: glyph, size: icon, color: theme.ink)
                Text(label).font(.system(size: font, weight: .medium)).foregroundStyle(theme.ink).lineSpacing(2).multilineTextAlignment(.center).lineLimit(2)
            }
            .frame(maxWidth: .infinity).padding(.horizontal, 6).padding(.vertical, 14)
        }
        .buttonStyle(.plain)
        .frame(maxWidth: .infinity)
    }
}

/// 收藏 / 复制后的轻提示（RN ReadVerseBookmarkFeedback）：底部 108 + 安全区，深色胶囊，进 160ms · 停 1400ms · 出 220ms
struct VerseFeedbackToast: View {
    let message: String?

    var body: some View {
        if let message {
            Text(message)
                .font(.system(size: 14, weight: .medium))
                .foregroundStyle(Color(rgb: 0xF7F4EF))
                .multilineTextAlignment(.center)
                .padding(.horizontal, 16).padding(.vertical, 10)
                .background(Capsule().fill(Color(red: 28 / 255, green: 20 / 255, blue: 16 / 255, opacity: 0.82)))
                .transition(.opacity)
                .allowsHitTesting(false)
        }
    }
}

/// 复制 / 分享用的经文文案，与 RN 同一格式：
/// 复制 = 「书名 章:节 经文」一行（lib formatScriptureVerseClipboard）；分享 = 「书名 章:节」换行经文（Share.share message）
enum VerseShareText {
    static func clipboard(bookName: String, chapter: Int, verse: Int, text: String) -> String {
        "\(bookName) \(chapter):\(verse) \(text.trimmingCharacters(in: .whitespacesAndNewlines))".trimmingCharacters(in: .whitespacesAndNewlines)
    }
    static func share(bookName: String, chapter: Int, verse: Int, text: String) -> String {
        "\(bookName) \(chapter):\(verse)\n\(text)"
    }
}
